### PR TITLE
ENH: Add weights option to numpy.quantile

### PR DESCRIPTION
**Pull request by iii-org-tw**
* We add a new feature to  numpy.quantile and numpy.percentile.
* Now, user can specify the weights of samples when computing quantiles.
* The reference is: https://stats.stackexchange.com/questions/13169/defining-quantiles-over-a-weighted-sample?fbclid=IwAR07lNSQsmd40zMB_DAgTdFLxpDoLKlv9QdrvoexSATfiQ_BdU_4aU9YtLI

####
* w must have the same shape with a.shape or a[axis] when axis is int or None.

* When w.shape == a.shape, it's the weights of each samples in a.

* If w.shape != a.shape, it's the weights for each dimensions in a.

  For example, 

            a = np.array([[1,2,3],[4,5,6]])

            w = np.array([10,11,12])

            axis = 1

  both weights of [1,2,3] and [4,5,6] are [10,11,12]

### CODE REVIEW

**Comment by eric-wieser** [comment]
This was proposed before in #9211. The difficulty came in exactly defining how to apply the weights.

**Comment by iii-org-tw** [comment]
@eric-wieser 
We have seen at least three types of formula to compute weighted quantiles.
SAS: http://support.sas.com/documentation/cdl/en/procstat/68142/HTML/default/viewer.htm#procstat_univariate_details14.htm
Wiki: https://en.wikipedia.org/wiki/Percentile
Stackexchange: https://stats.stackexchange.com/questions/13169/defining-quantiles-over-a-weighted-sample?fbclid=IwAR07lNSQsmd40zMB_DAgTdFLxpDoLKlv9QdrvoexSATfiQ_BdU_4aU9YtLI
We implement the third formula for the following reasons:
1. The same formula is adapted by R wtd quantile. (https://www.rdocumentation.org/packages/reldist/versions/1.6-6/topics/wtd.quantile)
2. This formula is well defined and can produce the same result as np.quantile when all value share the same weight.
3. As current np.quantile has different interpolation methods, we see that this formula can satisfy people's need as long as we clarify how we calculate weighted quantile.

**Comment by charris** [comment]
You have picked up a bunch of extraneous merges from master (19 commits), might want to rebase on master.

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -3655,16 +3655,17 @@ def _median(a, axis=None, out=None, overwrite_input=False):
...
 
 @array_function_dispatch(_percentile_dispatcher)
 def percentile(a, q, axis=None, out=None,
-               overwrite_input=False, interpolation='linear', keepdims=False):
+               overwrite_input=False, interpolation='linear',
+               keepdims=False, w=None):
     """
     Compute the q-th percentile of the data along the specified axis.
-
+
```
Comment (playma) [inline numpy/lib/function_base.py:3668]: Remove useless indent

File: numpy/lib/function_base.py [hunk 2]
```diff
@@ -3954,6 +4082,175 @@ def _lerp(a, b, t, out=None):
...
+            t[~unequal_indices] = 0
+            r = _lerp(x_below, x_above, t, out=out)
+
+    # If any slice contained a nan, then all results on that slice are also nan
+    if np.any(n):
+        r[..., n] = a.dtype.type(np.nan)
+        
+    recover_index_q = np.argsort(sorted_index_q)
+    r = r[recover_index_q]
+    if d == 0:
```
Comment (playma) [inline numpy/lib/function_base.py:4248]: return r[0] if d == 0 else r

File: numpy/lib/function_base.py [hunk 3]
```diff
@@ -3803,21 +3812,79 @@ def percentile(a, q, axis=None, out=None,
...
+
+@array_function_dispatch(quantile_dispatcher)
 def quantile(a, q, axis=None, out=None,
-             overwrite_input=False, interpolation='linear', keepdims=False):
+             overwrite_input=False, interpolation='linear',
+             keepdims=False, w=None):
     """
-    Compute the q-th quantile of the data along the specified axis.
-
-    .. versionadded:: 1.15.0
```
Comment (playma) [inline numpy/lib/function_base.py:3820]: Irrelevant code should not be deleted.

File: numpy/lib/function_base.py [hunk 4]
```diff
@@ -3912,37 +3991,86 @@ def quantile(a, q, axis=None, out=None,
...
+                        "Length of weights not compatible with a's size"
+                    )
+            else:
+                if w.shape[0] != a.shape[axis]:
+                    raise ValueError(
+                        "Length of weights not compatible with specified axis."
+                    )
+                w = np.broadcast_to(w, (a.ndim-1) * (1,) + w.shape)
+                w = w.swapaxes(-1, axis) * np.ones(a.shape)
+        if not _weight_is_valid(w):
```
Comment (playma) [inline numpy/lib/function_base.py:4019]: Add a new line above.

File: numpy/lib/function_base.py [hunk 5]
```diff
@@ -3912,37 +3991,86 @@ def quantile(a, q, axis=None, out=None,
...
-def _quantile_unchecked(a, q, axis=None, out=None, overwrite_input=False,
-                        interpolation='linear', keepdims=False):
+def _quantile_unchecked(a, q, w=None, axis=None, out=None,
+                        overwrite_input=False, interpolation='linear',
+                        keepdims=False):
     """Assumes that q is in [0, 1], and is an ndarray"""
-    r, k = _ureduce(a, func=_quantile_ureduce_func, q=q, axis=axis, out=out,
-                    overwrite_input=overwrite_input,
-                    interpolation=interpolation)
+    if w is None:
```
Comment (playma) [inline numpy/lib/function_base.py:4030]: It works fine when `w is None` and `w is not None`

```
r, k = _weighted_ureduce(a, func=_weighted_quantile_ureduce_func, w=w,
    q=q, axis=axis, out=out,
    overwrite_input=overwrite_input,
    interpolation=interpolation)
```

File: numpy/lib/function_base.py [hunk 6]
```diff
@@ -3912,37 +3991,86 @@ def quantile(a, q, axis=None, out=None,
...
+                                 interpolation=interpolation)
     if keepdims:
         return r.reshape(q.shape + k)
     else:
         return r
 
 
 def _quantile_is_valid(q):
-    # avoid expensive reductions, relevant for arrays with < O(1000) elements
+    # Avoid expensive reductions, relevant for arrays with < O(1000) elements
```
Comment (playma) [inline numpy/lib/function_base.py:4046]: Irrelevant code should not be changed.

File: numpy/lib/function_base.py [hunk 7]
```diff
@@ -3912,37 +3991,86 @@ def quantile(a, q, axis=None, out=None,
...
     if q.ndim == 1 and q.size < 10:
         for i in range(q.size):
             if q[i] < 0.0 or q[i] > 1.0:
                 return False
     else:
-        # faster than any()
+        # Faster than any()
```
Comment (playma) [inline numpy/lib/function_base.py:4052]: Irrelevant code should not be changed.

File: numpy/lib/function_base.py [hunk 8]
```diff
@@ -3954,6 +4082,175 @@ def _lerp(a, b, t, out=None):
     return lerp_interpolation
 
 
+def _find_weighted_index(sk, qsn, interpolation='linear'):
+    dim = sk.shape # (N, d1, d2,..., dk)
+    Nx = dim[0]
```
Comment (playma) [inline numpy/lib/function_base.py:4087]: Use snake case. ex: `nx`

File: numpy/lib/function_base.py [hunk 9]
```diff
@@ -3954,6 +4082,175 @@ def _lerp(a, b, t, out=None):
...
+    if Nx == 1:
+        return np.zeros_like(qsn)
+    
+    _sk = sk.reshape(dim[0], -1) # (N,-1)
+    _qsn = qsn.reshape(qsn.shape[0], -1) # (q,-1)
+    indices = []
+    
+    for j in range(_qsn.shape[1]):
```
Comment (playma) [inline numpy/lib/function_base.py:4095]: `i` for the first level, `j` for the second level.

File: numpy/lib/function_base.py (hunk 3 above)
Comment (r08922129) [reply numpy/lib/function_base.py:3820]: I fetched code with older version at first and caused that deletion.
This section is not in the master.

File: numpy/lib/function_base.py (hunk 6 above)
Comment (r08922129) [reply numpy/lib/function_base.py:4046]: Following PEP 8,  the first word in comments should be capitalized.
I confused whether to modify it.

File: numpy/lib/function_base.py (hunk 8 above)
Comment (r08922129) [reply numpy/lib/function_base.py:4087]: The `_quantile_ureduce_func` which computes non-weighted quantiles terms sample size Nx.
Maybe it's okay to keep them with the same name?

File: numpy/lib/function_base.py (hunk 3 above)
Comment (eric-wieser) [reply numpy/lib/function_base.py:3820]: @r08922129, it definitely is in master: 

https://github.com/numpy/numpy/blob/a0028bca0117874606bce99261d978df8d3f6610/numpy/lib/function_base.py#L3818-L3820

Comment (r08922129) [reply numpy/lib/function_base.py:3820]: Sorry, I'll check it again

File: numpy/lib/function_base.py (hunk 6 above)
Comment (playma) [reply numpy/lib/function_base.py:4049]: It's better to create a new PR to revise it. Not in this PR.

**Comment by iii-org-tw** [comment]
Hi @eric-wieser 

What can we do to help getting this PR accepted?

**Review by eric-wieser — CHANGES_REQUESTED** [review]
To move forward, I think we need to first come to a conclusion about gh-10736

File: numpy/lib/function_base.py (hunk 6 above)
Comment (mattip) [reply numpy/lib/function_base.py:4049]: It's fine to do this here and there in this PR, as long as there are not too many changes like this.

File: numpy/lib/function_base.py (hunk 9 above)
Comment (mattip) [reply numpy/lib/function_base.py:4098]: why?

**Review by mattip — COMMENTED** [review]
Let's discuss it at this week's triage meeting

File: numpy/lib/function_base.py (hunk 9 above)
Comment (playma) [reply numpy/lib/function_base.py:4098]: Common usage?

Comment (iii-org-tw) [reply numpy/lib/function_base.py:4098]: It doesn't matter here.
Also, to iterate over A[i,j], j should be put on the outer loop and A[i,j] is common usage.

**Comment by charris** [comment]
close/reopen

**Comment by charris** [comment]
There are a lot of lint errors that need fixing. Hit details to see them.  @mattip, do you recall the triage results?

**Comment by mattip** [comment]
No I don't remember and the [meeting notes](https://github.com/numpy/archive/blob/main/triage_meetings/triage-2020-08-26.md) do not mention it. I think though that @WarrenWeckesser self-assigned this?

**Comment by seberg** [comment]
We discussed this in the Triage meeting a bit, and settled on "we want this".  We are OK with the sorting approach and limiting e.g. only to `aweights`, that may need settling 100%, but generally using `aweights` (and maybe `fweights`) sems good.

**Comment by iii-org-tw** [comment]
> We discussed this in the Triage meeting a bit, and settled on "we want this". We are OK with the sorting approach and limiting e.g. only to `aweights`, that may need settling 100%, but generally using `aweights` (and maybe `fweights`) sems good.

So what should we do to make it merged?
Only to resolve the conflicts?

**Comment by iii-org-tw** [comment]
@seberg We have finished rebasing but there are still some problems. 

Now we just add `aweights` to `nanquantile` and `nanpercentile` to pass the test (they should have the same signature with `quantile` and `percentile`). As `nanquantile` removes nan by `_remove_nan_1d(arr1d)`,  elements in `aweights` should also be removed if the corresponding values in arr1d will be removed. Adding `aweights` to `_remove_nan_1d(arr1d)` as a new paramter seems not a good idea as not all functions will use `aweights`. However, writing another function makes duplicated code.

**Comment by iii-org-tw** [comment]
I run tests locally and got no error but after I push the code, the Refguide Check failed.
Here is the details.
```
2022-03-20T19:57:01.3135540Z ##[section]Starting: Run Refguide Check
2022-03-20T19:57:01.3145860Z ==============================================================================
2022-03-20T19:57:01.3146570Z Task         : Command line
2022-03-20T19:57:01.3147240Z Description  : Run a command line script using Bash on Linux and macOS and cmd.exe on Windows
2022-03-20T19:57:01.3147780Z Version      : 2.200.2
2022-03-20T19:57:01.3148240Z Author       : Microsoft Corporation
2022-03-20T19:57:01.3148760Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/command-line
2022-03-20T19:57:01.3149360Z ==============================================================================
2022-03-20T19:57:01.4919970Z Generating script.
2022-03-20T19:57:01.4933150Z Script contents:
2022-03-20T19:57:01.4934200Z python runtests.py -g --refguide-check
2022-03-20T19:57:01.4935230Z ========================== Starting Command Output ===========================
2022-03-20T19:57:01.4968100Z [command]/bin/bash --noprofile --norc /Users/runner/work/_temp/47852ec2-6c7f-4cc8-9f21-1c0342ab7f28.sh
2022-03-20T19:58:10.0310370Z numpy.core ............................................................................................................................................................................................................................................................................................................................................................Warning: overflow encountered in short_scalars
2022-03-20T19:58:16.1473440Z .........................
2022-03-20T19:58:16.1552920Z numpy.f2py .....
2022-03-20T19:58:16.8775710Z numpy.linalg ............................................
2022-03-20T19:58:18.0608180Z numpy.lib ......................................................................F...................................................................................................
2022-03-20T19:58:18.1139120Z numpy.lib.recfunctions .......................
2022-03-20T19:58:18.2818590Z numpy.fft ....................
2022-03-20T19:58:19.3505970Z numpy.ma .................................................................................................................................................................................................................................
2022-03-20T19:58:19.5106950Z numpy.polynomial .........
2022-03-20T19:58:19.5507870Z numpy.matrixlib ......
2022-03-20T19:58:58.1827610Z numpy.random .........................................................
2022-03-20T19:58:58.5037610Z numpy.testing ...........................................
2022-03-20T19:58:58.5041000Z Running checks for 11 modules:
2022-03-20T19:58:58.5042070Z 
2022-03-20T19:58:58.5042620Z =========
2022-03-20T19:58:58.5044680Z numpy.lib
2022-03-20T19:58:58.5045170Z =========
2022-03-20T19:58:58.5045450Z 
2022-03-20T19:58:58.5045850Z numpy.lib.quantile
2022-03-20T19:58:58.5047140Z ------------------
2022-03-20T19:58:58.5047440Z 
2022-03-20T19:58:58.5048440Z File "build/testenv/lib/python3.8/site-packages/numpy/__init__.py", line 375, in quantile
2022-03-20T19:58:58.5049070Z Failed example:
2022-03-20T19:58:58.5049560Z     np.quantile(a, 0.5, aweights=w, axis=0)
2022-03-20T19:58:58.5050020Z Exception raised:
2022-03-20T19:58:58.5050480Z     Traceback (most recent call last):
2022-03-20T19:58:58.5051080Z       File "/Users/runner/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/doctest.py", line 1336, in __run
2022-03-20T19:58:58.5051950Z         exec(compile(example.source, filename, "single",
2022-03-20T19:58:58.5052570Z       File "<doctest quantile[7]>", line 1, in <module>
2022-03-20T19:58:58.5053140Z         np.quantile(a, 0.5, aweights=w, axis=0)
2022-03-20T19:58:58.5053750Z       File "<__array_function__ internals>", line 180, in quantile
2022-03-20T19:58:58.5055040Z       File "/Users/runner/work/1/s/build/testenv/lib/python3.8/site-packages/numpy/lib/function_base.py", line 4479, in quantile
2022-03-20T19:58:58.5055820Z         return _quantile_unchecked(
2022-03-20T19:58:58.5057020Z       File "/Users/runner/work/1/s/build/testenv/lib/python3.8/site-packages/numpy/lib/function_base.py", line 4501, in _quantile_unchecked
2022-03-20T19:58:58.5059150Z         r, k = _weighted_ureduce(a,
2022-03-20T19:58:58.5060370Z       File "/Users/runner/work/1/s/build/testenv/lib/python3.8/site-packages/numpy/lib/function_base.py", line 4205, in _weighted_ureduce
2022-03-20T19:58:58.5061170Z         r = func(a, aweights, **kwargs)
2022-03-20T19:58:58.5062380Z       File "/Users/runner/work/1/s/build/testenv/lib/python3.8/site-packages/numpy/lib/function_base.py", line 4737, in _weighted_quantile_ureduce_func
2022-03-20T19:58:58.5063310Z         wp = np.take_along_axis(wp, sorted_index, axis=0)
2022-03-20T19:58:58.5063960Z       File "<__array_function__ internals>", line 180, in take_along_axis
2022-03-20T19:58:58.5065270Z       File "/Users/runner/work/1/s/build/testenv/lib/python3.8/site-packages/numpy/lib/shape_base.py", line 170, in take_along_axis
2022-03-20T19:58:58.5066230Z         return arr[_make_along_axis_idx(arr_shape, indices, axis)]
2022-03-20T19:58:58.5067480Z       File "/Users/runner/work/1/s/build/testenv/lib/python3.8/site-packages/numpy/lib/shape_base.py", line 34, in _make_along_axis_idx
2022-03-20T19:58:58.5068400Z         raise ValueError(
2022-03-20T19:58:58.5068980Z     ValueError: `indices` and `arr` must have the same number of dimensions
2022-03-20T19:58:58.5069380Z 
2022-03-20T19:58:58.5069620Z 
2022-03-20T19:58:58.5070080Z ERROR:  failed checking numpy.lib
2022-03-20T19:58:58.6500040Z ##[error]Bash exited with code '1'.
2022-03-20T19:58:58.6519370Z ##[section]Finishing: Run Refguide Check
```
I don't know what's wrong because I can get the right result when I run `np.quantile(a, 0.5, aweights=w, axis=0)` locally.

**Comment by seberg** [comment]
@iii-org-tw I don't immediately remember how to run the refguide check locally best.  But, the example uses `a = np.array([[10, 7, 4], [3, 2, 1]])` and you have `w = np.ones(2)`.  I get the same error locally, and I think it is probably correct.

For the weights to broadcast with `a`, they should be `np.ones(3)` or `np.ones((2, 1))`?

**Comment by iii-org-tw** [comment]
@seberg
yup, thank you, I did not notice that `w` was re-assigned with another value. The docstring was wrong and have been modified. 

Now, I have rebased the branch and solved conflicts. A problem is now we add `aweights` to `nanquantile` and `nanpercentile` **JUST FOR PASSING THE TESTS** (they should have the same signature with `quantile` and `percentile`). As `nanquantile` removes nan by `_remove_nan_1d(arr1d)`, elements in `aweights` should also be removed if the corresponding values in `arr1d` will be removed. Adding `aweights` to `_remove_nan_1d(arr1d)` as a new paramter seems not a good idea since not all functions will use `aweights`. However, writing another function makes duplicated code. What do you think?

**Comment by seberg** [comment]
Closing, gh-24254 will probably happen, although I am sure it is *not* the same as this was proposing.

**Pull request closed without merging**